pkg/util/deadlock-detector_test.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"sync"
+	"testing"
+	"time"
+)
+
+type fakeExiter struct {
+	format string
+	args   []string
+	exited bool
+}
+
+func (f *fakeExiter) Exitf(format string, args ...string) {
+	f.format = format
+	f.args = args
+	f.exited = true
+}
+
+func TestDeadlockWatchdogLocked(t *testing.T) {
+	lock := &sync.RWMutex{}
+	lock.Lock()
+
+	exitCh := make(chan time.Time, 1)
+	fake := fakeExiter{}
+
+	detector := &deadlockDetector{
+		lock:          &rwMutexToLockableAdapter{lock},
+		name:          "test deadlock",
+		exitChannelFn: func() <-chan time.Time { return exitCh },
+		exiter:        &fake,
+	}
+
+	exitCh <- time.Time{}
+
+	detector.run()
+
+	if !fake.exited {
+		t.Errorf("expected to have exited")
+	}
+}
+
+func TestDeadlockWatchdogUnlocked(t *testing.T) {
+	lock := &sync.RWMutex{}
+
+	fake := fakeExiter{}
+
+	detector := &deadlockDetector{
+		lock:          &rwMutexToLockableAdapter{lock},
+		name:          "test deadlock",
+		exitChannelFn: func() <-chan time.Time { return time.After(time.Millisecond * 500) },
+		exiter:        &fake,
+	}
+
+	// TODO: this is not a great way to write this test, but I can't think of a better way
+	go func() {
+		time.Sleep(2 * time.Second)
+		detector.stop = true
+	}()
+
+	detector.run()
+
+	if fake.exited {
+		t.Errorf("expected to have not exited")
+	}
+}
+
+func TestDeadlockWatchdogLocking(t *testing.T) {
+	lock := &sync.RWMutex{}
+
+	fake := fakeExiter{}
+
+	go func() {
+		for {
+			lock.Lock()
+			lock.Unlock()
+		}
+	}()
+
+	detector := &deadlockDetector{
+		lock:          &rwMutexToLockableAdapter{lock},
+		name:          "test deadlock",
+		exitChannelFn: func() <-chan time.Time { return time.After(time.Millisecond * 500) },
+		exiter:        &fake,
+	}
+
+	// TODO: this is not a great way to write this test, but I can't think of a better way
```

Comment:
maybe split run into run and runOnce, and call the latter in tests a fixed number of times?